### PR TITLE
Fix refprune on obfuscated refs and stabilize optimisation WRT wrappers.

### PR DESCRIPTION
**Pull request by stuartarchibald**
Also makes more things vectorize. As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This needs an updated llvmlite.

**Comment by stuartarchibald** [comment]
Depends on https://github.com/numba/llvmlite/pull/647

File: numba/core/callconv.py [hunk 1]
```diff
@@ -561,7 +568,18 @@ def call_function(self, builder, callee, resty, argtys, args):
...
         realargs = [retvaltmp, excinfoptr] + args
-        code = builder.call(callee, realargs)
+        # deal with attrs, it's find to specify a load in a string like
+        # "noinline fast" as per LLVM or equally as an iterable of individual
+        # attributes.
+        if attrs is None:
+            _attrs = ()
+        elif isinstance(attrs, str):
+            _attrs = tuple(attrs.split(' '))
+        elif isinstance(attrs, Iterable):
```
Comment (sklam) [inline numba/core/callconv.py:578]: This is never tested.

Comment (stuartarchibald) [reply numba/core/callconv.py:578]: I can remove, though it just feels a bit strange API wise writing:
```
call_function(self, builder, callee, resty, argtys, args, attrs="noinline nounwind"):
```
vs.
```
call_function(self, builder, callee, resty, argtys, args, attrs=["noinline", "nounwind"]):
```
but I think testing the latter might be hard.

Comment (sklam) [reply numba/core/callconv.py:578]: Agree that the latter is nicer. And, I'd rather we have only one way of spelling it, even if we have to say `attrs=['noinline']`

**Comment by sklam** [comment]
Need to bump llvmlite upper limit

**Comment by sklam** [comment]
Some tests are not vectorising again

File: numba/core/callconv.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/callconv.py:578]: I've made a change along these lines, see what you think.

**Comment by sklam** [comment]
Looks like whack-a-mole is done

**Review by sklam — APPROVED** [review]
Great that we will get back the RC1 speed.

**Pull request merged**